backend/drm/drm.c

```diff
@@ -208,10 +209,41 @@ static bool init_planes(struct wlr_drm_backend *drm) {
 		 * tracking of them for DRM lease clients.
 		 *
 		 * possible_crtcs is a bitmask of crtcs, where each bit is an
-		 * index into drmModeRes.crtcs. So if bit 0 is set (ffs starts
-		 * counting from 1), crtc 0 is possible.
-		 */
-		int crtc_bit = ffs(plane->possible_crtcs) - 1;
+		 * index into drmModeRes.crtcs. So if bit i is set crtc i is
+		 * possible.
+                 *
+		 * The loop below is selecting the first possible CRTC without a
+		 * associated plane of the considered type or, as a fallback,
+		 * the last possible CRTC.
+                 */
+
+		int crtc_bit ;
+		int num_crtcs = (int)drm->num_crtcs ;
+		assert(num_crtcs<=64);
+		uint64_t bits = plane->possible_crtcs ;
+		// Some drm implementations (e.g. vc4 on raspberry pi) may set more
+		// bits than there are CRTCs so the unused upper bits must be cleared
+		// to insure a proper detection of the last possible CRTC.
```

Comment:
Can you post the output of [drm_info](https://github.com/ascent12/drm_info) here?
That sounds like a kernel bug if this is true, and I hate adding workarounds for bugs.

Reply:
Here it is.

Recent 5.x kernels for Raspberry Pi are always producing CRTCs: {0, 1, 2, 3, 4, 5, 6, 7} for all planes regardless of the actual number of CRTCs:   
https://drmdb.emersion.fr/devices/01cd892b4221

Older 4.9 kernels (i.e. the kernel officially supported by the Raspberry Pi OS) are not setting unused bits and so produce CRTCs: {0,1} with 2 monitors.  
https://drmdb.emersion.fr/devices/e973c8924b7e

However, I just looked at the kernel driver code and I discovered that this is caused by the hard-coded 0xff here
 https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/gpu/drm/vc4/vc4_firmware_kms.c#L820

This is a modified kernel provided by the Raspberry Pi foundation. As far as I can tell, the vc4 driver in the mainline Linux kernels is not behaving like that.  Even worse, the association of the planes to the CRTCs is also hardcoded meaning that a  `possible_crtcs` with more than one bit is wrong. 

For example, according to the current `possible_crtcs` it should be legal to assign the 1st cursor plane to the 2nd crtc and the 2nd cursor plane to the 1st crtc. I just tried that in sway and the effect was that my left screen was showing the right screen cursor and the other way around. So this is definitely a bug and I will fill a bug report in the rpi kernel. 

Though, it would be nice if my pull request was accepted to allow rpi4 users to use multiple screens without having to wait for a new kernel. A basic plane matching logic is better than nothing. 

Also clearing the unused upper bits is not really needed even for the rpi4 because the driver is allocating exactly one plane of each type for each crtc so the fallback is never needed. However, that could become a problem if the rpi4 driver was modified to allow more than one overlay per crtc.

Reply:
If your main concern is the cleaning of the upper bits, I can rewrite the loop as follow. The behavior should be identical with the last valid bit being used as a fallback.

    int num_crtcs = (int)drm->num_crtcs ;
    assert(num_crtcs<=64);
    int crtc_bit = -1;
    uint64_t bits = plane->possible_crtcs ;
    for (int j=0;j<num_crtcs;j++) {
       if (bits & 1) {
          crtc_bit = j;
          struct wlr_drm_crtc *crtc = &drm->crtcs[crtc_bit];
          if (type == DRM_PLANE_TYPE_PRIMARY && crtc->primary) {
             continue;
          }
          if (type == DRM_PLANE_TYPE_CURSOR && crtc->cursor) {
             continue;
          }
          if (type == DRM_PLANE_TYPE_OVERLAY && crtc->num_overlays>0) {
             continue;
          }
          break;
       }
       bits >>= 1;                  
    }

Reply:
>However, I just looked at the kernel driver code and I discovered that this is caused by the hard-coded 0xff here

That's just straight up broken.

>Though, it would be nice if my pull request was accepted to allow rpi4 users to use multiple screens without having to wait for a new kernel

I absolutely despise driver-specific code, and driver-specific workarounds is even worse. Once you add them in, they become extremely hard to remove, and just leads to a maintenance burden. The case is even worse when we're not talking about an upstream kernel.

>However, that could become a problem if the rpi4 driver was modified to allow more than one overlay per crtc.

I believe it's kernel UAPI[1] to have exactly 1 primary plane per CRTC, meaning the current `ffs()` algorithm works fine.
I'm not quite as sure if it's UAPI for cursor planes, but it doesn't actually make sense to have more than 1 cursor plane anyway; it's kind of a legacy idea for userspace's convenience, and any sane driver writer would make that an overlay plane instead.

Overlay planes currently are not important. We track them (because we kind of have to), but we don't actually use them. The comment in the code explains that the algorithm we use is intentionally crap, and we're willing to accept bad results. Once we're doing planes properly with libliftoff, then a real matching algorithm needs to be used (which can't reliably be done ahead of time), but libliftoff itself handles that.

---
[1] https://01.org/linuxgraphics/gfx-docs/drm/gpu/drm-kms.html#plane-abstraction : "All drivers should provide one primary plane per CRTC to avoid surprising userspace too much"

Reply:
I assume that this means that the pull request will not be accepted. I do not really mind. My patched wlroots is working fine so I will simply attach a patch to #1943. That should be enough to please the few raspberry pi users concerned with that problem.

Reply:
@ascent12 I asked on #dri-devel and they said it's not a driver bug (https://github.com/swaywm/wlroots/issues/1943#issuecomment-561261556).

Reply:
@emersion It depends what you asked for. I agree than setting bit for non-existing CRTCs is not really a bug. Also, setting multiple CRTCs for each plane is legal and that could be an argument for my pull request.. 

However in that specific case, I do not believe anymore that the Raspberry Pi kernel implementation is correct because a quick look at the `vc4_fkms_create_screen` function (https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/gpu/drm/vc4/vc4_firmware_kms.c#L1738) shows that they are manually creating and associating 3 planes to each CRTC. As I explained previously, swapping the association order of the cursor planes results in the cursors appearing on the wrong screen. This is definitely not the expected behavior.

Reply:
https://lists.freedesktop.org/archives/dri-devel/2020-August/275142.html